com.avaloq.tools.ddk.xtext/src/com/avaloq/tools/ddk/xtext/modelinference/InferredModelAssociator.java

```diff
@@ -214,7 +230,70 @@ public void discardDerivedState(final DerivedStateAwareResource resource) {
       derived.add(eObject);
     }
     resourcesContentsList.removeAll(derived);
-    getSourceToInferredModelMap(resource).clear();
+    getSourceToTargetMap(resource).clear();
+  }
+
+  private static <K, V> void putIntoSmallSetMap(final K key, final V value, final Map<? super K, Set<V>> map, final boolean head) {
+    Set<V> set = map.get(key);
+    if (head) {
+      if (set == null) {
+        set = new LinkedListBasedSet<V>();
+        map.put(key, set);
+      }
+      set.remove(value);
+      ((LinkedListBasedSet<V>) set).elements.addFirst(value);
+    } else {
+      if (set == null) {
+        set = new LinkedListBasedSet<V>();
+        map.put(key, set);
+      }
+      set.add(value);
+    }
+  }
+
+  /**
+   * Set implementation based on {@link LinkedList}.
+   *
+   * @param <E>
+   *          generic element type
+   */
+  public static class LinkedListBasedSet<E> extends AbstractSet<E> {

```

Comment:
can you pls explain why this special set is necessary

Reply:
This is needed because we want to have an ordered set (at least for the values of targetToSourceMap), where we know that the first